mesop/runtime/context.py

```diff
@@ -120,9 +125,36 @@ def serialize_state(self) -> pb.States:
       states.states.append(pb.State(data=serialize_dataclass(state)))
     return states
 
+  def diff_state(self) -> pb.States:
+    states = pb.States()
+    for (key, state), previous_state in zip(
+      self._states.items(), self._previous_states.values()
+    ):
+      states.states.append(pb.State(data=diff_state(previous_state, state)))
+      # If state has not changed no need to update the previous state since it should be
+      # the same.
+      #
+      # One thing to note is that there seems to be an issue multiple states with the
+      # same key can exist (i.e. usage in Mesop demo app). It's unclear if this causes
+      # problem here.
+      if state != self._previous_states[key]:
+        self._previous_states[key] = copy.deepcopy(state)
+    return states
+
   def update_state(self, states: pb.States) -> None:
-    for state, proto_state in zip(self._states.values(), states.states):
+    for state, previous_state, proto_state in zip(
+      self._states.values(), self._previous_states.values(), states.states
+    ):
       update_dataclass_from_json(state, proto_state.data)
+      # We should check if state and previous state are the same. There should be no
```

Comment:
Same as above, I don't understand why this is needed.

Reply:
This case is still needed for now since we need to update the previous state to match the state at the start of each ui request since the state gets reset.

Reply:
To clarify, I mean that for each ui request, you should have a new context, so previous_states should get set in `__init__` so you shouldn't need to update the previous state.

Reply:
Yes, it's true that a new context gets created. But the previous states will equal the default state when the state is initialized. But we need the previous state to be the state at the start of the request. So in the case of demo or an app with many pages, most should have the default previous state. But the current page will need to the state updated. So there's some optimization that we can do. However we'd probably need to resolve #391 first.

Reply:
ok yeah, you're right